CIPs/cip-37.md

```diff
@@ -14,42 +14,273 @@ created: 2021-01-07
+
+Importantly, the optional `address-type` is **NOT** part of the checksum computation.
+
+The 40-bit number returned by PolyMod is split into eight 5-bit numbers (msb first).
+The payload and the checksum are then encoded according to the base32 character table.
+
+5. Concatenate the following parts to get the final address: `[network-prefix]`, `":"`, `[payload]`, `[checksum]`
+
+<!-------------------->
+
+### Decoding
+
+INPUT: a conflux-base32-address
+
+OUTPUT: a conflux-hex-address (20 bytes) and a network-id (4 bytes)
+
+1. Treating the address as an ASCII-string, split at `":"`. If the resulting array has less than 2 items, reject. We treat the first item as `network-prefix-raw`, and the last item as `payload-raw`. Any other items in-between are treated as optional key-value pairs.
+
+2. Parse **network-id**:
+
+```
+    match network-prefix-raw
+        case "cfx":     1029
+        case "cfxtest": 1
+        case "cfx1029": reject
+        case "cfx1":    reject
+        case "cfx[n]":  n if n fits into 4 bytes
+        otherwise:      reject
+```
+
+3. Validate payload:
+
+    - If the length of `payload-raw` is not `42`, reject.
+    - If `payload-raw` contains both lowercase and uppercase characters, reject.
+    - If `payload-raw` contains invalid characters (not in the alphabet), reject.
+
+4. Validate checksum.
+
+To verify a base32-formatted address, the input data (list of integers) for the PolyMod function is assembled from these parts:
+- The lower 5 bits of each character of the prefix.
+- A zero for the separator (5 zero bits).
+- Each base32 char of the payload mapped to its respective number.
+
+If PolyMod returns non-zero, then the address was broken, reject.
+
+5. Convert `payload-raw` into 8-bit array, split into `version-byte` and `payload`.
+
+6. Validate `version-byte`: If `version-byte` is **NOT** `0x00`, reject.
+
+7. Validate **address-type**  (if provided):
+
+    - If `payload[0] & 0xf0` is `b00000000` and `address-type` is **NOT** `"builtin"`: reject.
+    - If `payload[0] & 0xf0` is `b00010000` and `address-type` is **NOT** `"user"`: reject.
+    - If `payload[0] & 0xf0` is `b10000000` and `address-type` is **NOT** `"contract"`: reject.
+
+Return `(payload, network-id)`.
+
+*(Note: Conflux nodes are expected to validate `network-id` against their local chain id and reject addresses that do not match. This check is not part of this specification.)*
+
+<!-------------------->
+
+### Example
+
+```
+encode(0x106d49f8505410eb4e671d51f7d96d2c87807b09, 1029, true)
+
+1. network-prefix: 1029 => "cfx"
+2. address-type: "user"
+3. version-byte: 0x00
+4. payload: [0x00, 0x10, 0x6d, 0x49, 0xf8, 0x50, 0x54, 0x10, 0xeb, 0x4e, 0x67, 0x1d, 0x51, 0xf7, 0xd9, 0x6d, 0x2c, 0x87, 0x80, 0x7b, 0x09]
+   5-bit parts: [0x00, 0x00, 0x08, 0x06, 0x1a, 0x12, 0x0f, 0x18, 0xa, 0x1, 0xa, 0x1, 0x1, 0x1a, 0x1a, 0xe, 0xc, 0x1c, 0xe, 0x15, 0x03, 0x1d, 0x1e, 0x19, 0x0d, 0x14, 0x16, 0x08, 0x0f, 0x00, 0x03, 0x1b, 0x01, 0x04]
```

Comment:
Some of these are padded with zero on the left, some are not. Is this intentional?